symfony/security-bundle/7.3/config/packages/security.yaml

```
@@ -0,0 +1,39 @@
+security:
+    # https://symfony.com/doc/current/security.html#registering-the-user-hashing-passwords
+    password_hashers:
+        Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface: 'auto'
+    # https://symfony.com/doc/current/security.html#loading-the-user-the-user-provider
+    providers:
+        users_in_memory: { memory: null }
+    firewalls:
+        dev:
+            pattern: ^/_profiler|_wdt|assets|build/ # `assets` is for AssetMapper; `build` is for Webpack Encore
```

```suggestion
            pattern: ^/(_profiler|_wdt|assets|build)/ # `assets` is for AssetMapper; `build` is for Webpack Encore
```
Shouldn't this be with parentheses instead?

---

Not needed if in the union is the full pattern.

---

Yes, but here we have `^/` and `/` that are not part of the union right?

---

I think @mtarld is right; https://www.php.net/manual/en/regexp.reference.alternation.php doesn't say anything about the alternative branches stopping at special characters. So currently, the 4 alternatives are:
* `^/_profiler`
* `_wdt`
* `assets`
* `build/`

But it "works" nevertheless, since `_wdt` matches anything like `/_wdt/whatever...` ;-)

So the way to go IMO is to bring back the parentheses.

However, one thing is irritating me: While testing this, I commented out that line in my `security.php`, and the web debug toolbar is still showing (and all assets loading) in DEV environment.

---

You commented out that single line or the whole dev firewall?

---

Oh correct, parenthesis are required because it's not the full pattern. The `/.../` confused me into thinking that they were pattern delimiters 😅

---

@ThomasLandauer try having an access_control rule requiring some authentication for all pages except the login page. When accessing the login page, the toolbar would not load as it would suddenly require authentication. That's why we have this `dev` firewall disabling authentication for dev URLs.

---

OK, thanks, so this line is only needed if there's an `access_control`? And you're assuming that most people have one?
So what about adding some comment here, like:
> Exception for the web debug toolbar and assets (only needed if you have a broad `access_control`):

About the (missing) delimiter: This irritated me too. The best solution would probably be to include the delimiter here in the string. But that would be a breaking change, which is probably not worth it. But what about at least adding some comment like:
> The delimiter `xxx` is added automatically

This might be helpful if the delimiter happens to appear in somebody's path (cause they'd have to escape it then).
=> So which delimiter is it?

---

If you don't have any access control rule enforcing a broad requirement, the fact that the toolbar would be matched by the main firewall may not prevent it from loading.
However, the use case of securing the whole project under login (except a few page like the login and password recovery pages) is a common use case, and we want the toolbar and profiler to work out of the box (otherwise, things would be confusing for users). That's why we have this `dev` firewall disabling any authentication for the dev tooling pages.

> => So which delimiter is it?

the PathRequestMatcher uses `{}` delimiters, which don't require special escaping inside the regex (thanks to the balancing of braces, the engine can distinguish whether a `}` is closing a repetition quantifier or is the end delimiter of the regex)